internal/usde/usde.go

```diff
@@ -67,69 +68,79 @@ func resourceSpecNeedsUpdating(ctx context.Context, newDef *kubernetes.GenericOb
 	if err != nil {
 		return false, apiv1.UpgradeStrategyError, err
 	}
 
-	numaLogger.WithValues(
-		"newSpecOnlyApplyPaths", newSpecOnlyApplyPaths,
-		"newSpecWithoutApplyPaths", newSpecWithoutApplyPaths,
-	).Debug("split new spec")
+	existingDefUnstr, err := kubernetes.ObjectToUnstructured(existingDef)
+	if err != nil {
+		return false, apiv1.UpgradeStrategyError, err
+	}
 
-	// Split existingDef
-	existingSpecOnlyApplyPaths, existingSpecWithoutApplyPaths, err := util.SplitObject(existingDef.Spec.Raw, applyPaths, []string{}, ".")
+	upgradeStrategy, err := getDataLossUpggradeStrategy(ctx, newDef.Namespace)
 	if err != nil {
 		return false, apiv1.UpgradeStrategyError, err
 	}
 
 	numaLogger.WithValues(
-		"existingSpecOnlyApplyPaths", existingSpecOnlyApplyPaths,
-		"existingSpecWithoutApplyPaths", existingSpecWithoutApplyPaths,
-	).Debug("split existing spec")
+		"usdeConfig", usdeConfig,
+		"dataLossFields", dataLossFields,
+		"upgradeStrategy", upgradeStrategy,
+		"newDefUnstr", newDefUnstr,
+		"existingDefUnstr", existingDefUnstr,
+	).Debug("started deriving upgrade strategy")
+
+	// Loop through all the data loss fields from config to see if any changes based on those fields require a data loss prevention strategy
+	for _, dataLossField := range dataLossFields {
+		newDefField, newIsMap, err := util.ExtractPath(newDefUnstr.Object, strings.Split(dataLossField.Path, "."))
+		if err != nil {
+			return false, apiv1.UpgradeStrategyError, err
+		}
 
-	// Compare specs without the apply fields and check user's strategy to return their preferred strategy
-	if !reflect.DeepEqual(newSpecWithoutApplyPaths, existingSpecWithoutApplyPaths) {
-		upgradeStrategy, err := getDataLossUpggradeStrategy(ctx, newDef.Namespace)
+		existingDefField, existingIsMap, err := util.ExtractPath(existingDefUnstr.Object, strings.Split(dataLossField.Path, "."))
 		if err != nil {
 			return false, apiv1.UpgradeStrategyError, err
 		}
 
 		numaLogger.WithValues(
-			"upgradeStrategy", upgradeStrategy,
-			"newSpecWithoutApplyPaths", newSpecWithoutApplyPaths,
-			"existingSpecWithoutApplyPaths", existingSpecWithoutApplyPaths,
-		).Debug("the specs without the 'apply' paths are different")
-
-		return true, upgradeStrategy, nil
+			"dataLossField", dataLossField,
+			"newDefField", newDefField,
+			"existingDefField", existingDefField,
+			"newIsMap", newIsMap,
+			"existingIsMap", existingIsMap,
+		).Debug("checking data loss field differences")
+
+		areDefFieldsMap := newIsMap && existingIsMap
+
+		// If the current field is not a map or if it is (assumed from the config if IncludeSubfields is true) and the config
+		// says to include comparing the subfields, then compare the fields/maps and, if the fields/maps are different,
+		// a data loss prevention strategy is needed
+		if (dataLossField.IncludeSubfields || !areDefFieldsMap) && !reflect.DeepEqual(newDefField, existingDefField) {
+			return true, upgradeStrategy, nil
+		}
```

Comment:
I'm confused. Where are we looking for the case of `dataLossField.IncludeSubfields==false` with a map? For example, if we add the field `spec.vertices.source.generator` and it wasn't there before. We want this to return PPND/Progressive, right?

Reply:
If we want the behavior you described, then the IncludeSubfields for that path must be set to `true`. This is the assumption.

Reply:
No, we want this to result in "data loss":

before:
```
spec:
  vertices:
     source:
```
after:
```
spec:
  vertices:
     source:
       generator:
```

But we don't want this to result in "data loss":

before:
```
spec:
  vertices:
     source:
       generator:
          key: value1
```

after:
```
spec:
  vertices:
     source:
       generator:
          key: value2
```

That is why `spec.vertices.source.generator` has `IncludeSubfields-=false`

Reply:
spoke offline that the logic is likely correct but hard to read - I can merge this PR and more carefully review it after you modify that

Reply:
As we discussed I also added 2 more test cases.